### DISCUSSION TITLE
DVI output incorrect.

### DISCUSSION THREAD

**Discussion by t1jsh111** [bwoods/TeX--#9] [Q&A] [answered]
Consider for example the simple tex file:
test.tex with content: 
"
This is \TeX 
\end
"

DVI output generated by TeX-Live:
![image](https://user-images.githubusercontent.com/25765405/141658544-d01db595-4c3b-4983-8e43-103361d09b13.png)

DVI output generated by the program:
![image](https://user-images.githubusercontent.com/25765405/141658486-528162eb-1a2c-42e0-93b0-aaf8f83e9893.png)

These HEX 0D (corresponding to unsigned int 13) should not be there. For some reason when Font Def 1(DVI command 243)  gets executed the program adds this hexadecimal in the dvi, causing the DVI to be improper DVI output.

**Comment by bwoods** [comment]
It looks as though TeX-Live is emitting unix newlines here (`\x0a` → `\n`) where the base `tex.web` version emits DOS newlines (`\x0d\x0a` → `\r\n`).

I assume the dvi file renders fine?

**Comment by t1jsh111** [comment]
Thank you for your reply.

According to the specification, after the 243 unsigned int (FNT_DEF 1 which is F3 in the hex) there should be precisely 15 bytes before the the font description starts (the font description will be cmr10 in this case)
![image](https://user-images.githubusercontent.com/25765405/141659964-0026229d-6545-4171-bdbe-e5142081eb55.png)


Since in the dvi generated by the program there is this additional 0D twice, there will be 17 bytes instead before the font description starts. The DVI viewers expect it to start after 15 bytes, causing them not able to render these dvi files unfortunately (this is also how I detected this problem). I tested several dvi viewers but all were not able because of this problem.

I'm not sure why there should be a newline in s[4] and d[4] since these refer to scale factor and design size respectively. I suppose something went wrong here in the translation to c++ code.

Do you have any ideas about this?

It is much appreciated!

**Comment by bwoods** [comment, marked as answer]
Intersting. I haven't seen this problem. Are you running these tests on Windows? If so
-  perhaps the built in streams are “helpfully” doing some newline translations. If these are, and
-  if these are streams that you are passing into the library, make sure that they are opened in [binary mode](https://en.cppreference.com/w/cpp/io/ios_base/openmode).

**Comment by t1jsh111** [comment]
I was doing this on Windows indeed. Opening them in binary mode resolved the issue.
Sorry to have bothered you with an issue on my side.
Thank you for your help!
